File: C++/2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
        int carry = 0;
        ListNode dummy_root(0);
        ListNode *root = &dummy_root;
        while (l1 || l2 || carry) {
            carry = (l1 ? l1->val : 0) + (l2 ? l2->val : 0) + carry;
            root->next = new ListNode(carry % 10);
            carry = carry / 10;
            l1 = l1 ? l1->next : l1;
            l2 = l2 ? l2->next : l2;
            root = root->next;
        }
        return dummy_root.next;
    }
};

File: Python/2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        carry = 0
        root = ListNode(0)
        r = root
        while l1 != None or l2 != None:
            if l1:
                carry += l1.val
            if l2:
                carry += l2.val

            root.next = ListNode(carry % 10)
            carry //= 10
            root = root.next
            if l1 != None:
                l1 = l1.next
            if l2 != None:
                l2 = l2.next
        if carry == 1:
            root.next = ListNode(1)
        return r.next